dcm4che-core/src/main/java/org/dcm4che3/io/DicomInputStream.java

```diff
@@ -484,22 +561,41 @@ public Attributes readCommand() throws IOException {
     }
 
     public void readAllAttributes(Attributes attrs) throws IOException {
-        readAttributes(attrs, -1, o -> false);
+        readAttributes(attrs, UNDEFINED_LENGTH, o -> false);
     }
 
     public Attributes readDataset() throws IOException {
-        return readDataset(-1, o -> false);
+        return readDataset(o -> false);
     }
 
     public Attributes readDatasetUntilPixelData() throws IOException {
-        return readDataset(-1, o -> o.tag == Tag.PixelData);
+        return readDataset(o -> o.tag == Tag.PixelData);
     }
 
+    /**
+     * @deprecated Use one of the other {@link #readDataset()} methods instead. If you want to
+     * specify a length limit, you may supply a {@link LimitedInputStream} or use
+     * {@link #createWithLimit} or {@link #createWithLimitFromFileLength}.
+     */
     @Deprecated
     public Attributes readDataset(int len, int stopTag) throws IOException {
         return readDataset(len, tagEqualOrGreater(stopTag));
     }
 
+    public Attributes readDataset(int stopTag) throws IOException {
+        return readDataset(tagEqualOrGreater(stopTag));
+    }
+
+    public Attributes readDataset(Predicate<DicomInputStream> stopPredicate) throws IOException {
+        return readDataset(UNDEFINED_LENGTH, stopPredicate);
+    }
+
+    /**
+     * @deprecated Use one of the other {@link #readDataset()} methods instead. If you want to
+     * specify a length limit, you may supply a {@link LimitedInputStream} or use
+     * {@link #createWithLimit} or {@link #createWithLimitFromFileLength}.
+     */
+    @Deprecated
     public Attributes readDataset(int len, Predicate<DicomInputStream> stopPredicate) throws IOException {
```

Comment:
I guess the plan is the eventually make this method private?

Reply:
Or just keep it deprecated like this, as I don't think a library should break user's code, unless it is really needed.

But that is just my suggestion, let's see what @gunterze thinks.